Replaced SCMDescriptor#generation integer with AtomicInteger to fix the concurrency issue reported by SpotBugs

StefanSpieker opened the pull request:
Replaced integer with AtomicInteger to fix spotbugs issue [vo-volatile-increment](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#vo-an-increment-to-a-volatile-field-isn-t-atomic-vo-volatile-increment) in SCMS class

### Proposed changelog entries

* Fix the potential synchronization issue in SCMDescriptor#generation. The field was changed from public to private, and new API methods were added to allow accessing it

StefanSpieker committed:
replaced integer with AtomicInteger to fix spotbugs issue

daniel-beck commented on core/src/main/java/hudson/scm/SCMDescriptor.java:
```diff
@@ -56,7 +58,7 @@
      */
     @Deprecated
-    public volatile int generation = 1;
+    public volatile AtomicInteger generation = new AtomicInteger(1);
```
In theory this is public API available for use in plugins, and a breaking change if any of them do. Have you checked whether this is used in any plugins e.g. using https://github.com/jenkins-infra/usage-in-plugins ?

StefanSpieker replied:
Sorry, I should have mentioned, that I found it here: https://ci.jenkins.io/job/Infra/job/usage-in-plugins/job/master/lastSuccessfulBuild/artifact/output/deprecated-and-unused.html

daniel-beck replied:
In that case you should add `@Restricted(NoExternalUse.class) @RestrictedSince("TODO")` as well, to prevent future uses.

StefanSpieker replied:
Thanks, I added this.

StefanSpieker committed:
added @Restricted to prevent further usage

jvz commented on core/src/main/java/hudson/scm/SCMDescriptor.java:
```diff
@@ -56,7 +62,8 @@
     @Deprecated
-    public volatile int generation = 1;
+    @Restricted(NoExternalUse.class) @RestrictedSince("TODO")
+    public volatile AtomicInteger generation = new AtomicInteger(1);
```
Making a variable atomic and volatile is redundant. The AtomicInteger (or any other AtomicReference) can be final as the volatile field is contained in those classes.

StefanSpieker replied:
Thanks, I changed it.

StefanSpieker committed:
replaced volatile with final because of AtomicInteger usage

res0nance reviewed (approved):
LGTM

oleg-nenashev commented on core/src/main/java/hudson/scm/SCMDescriptor.java:
```diff
@@ -56,7 +62,8 @@
     @Deprecated
-    public volatile int generation = 1;
+    @Restricted(NoExternalUse.class) @RestrictedSince("TODO")
+    public final AtomicInteger generation = new AtomicInteger(1);
```
If we break compatibility anyway, why don't we make it private + add a proper getter method?

oleg-nenashev replied:
@StefanSpieker WDYT?

StefanSpieker replied:
Sorry I missed that. Yes I will change that.

oleg-nenashev commented on core/src/main/java/hudson/scm/SCMS.java:
```diff
@@ -60,7 +60,7 @@ public static SCM parseSCM(StaplerRequest req, AbstractProject target) throws Fo
             scm = new NullSCM(); // JENKINS-36043 workaround for AbstractMultiBranchProject.submit
         }
-        scm.getDescriptor().generation++;
+        scm.getDescriptor().generation.incrementAndGet();
```
no need to get it here

StefanSpieker replied:
I didn't want to change the behavior there and ++ is exaclty incrementAndGet(). And there is no method which just increments the value. So either getAndIncrement() or the chosen one.

oleg-nenashev replied:
OK, thanks

oleg-nenashev reviewed (commented):
I was unable to find any external usages, but it is still a breaking change in API. It must be referenced in the developer changelog at least

StefanSpieker committed:
implemented a getter and an increment method

StefanSpieker commented:
I reworked it with a getter and implemented a method for incrementing the value. I kept the deprecated warnings because it is just a refactoring to fix Spotbugs warnings and I guess it might be removed someday.

StefanSpieker committed:
fixed typeo in javadoc

oleg-nenashev reviewed (approved):
I think we can go ahead with that. It is a minor breaking change, but I was unable to find any usages.
I plan to merge it tomorrow if no negative feedback

reinholdfuereder commented:
Please have a look at https://issues.jenkins-ci.org/browse/JENKINS-60513
